scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,256 @@
+use crate::cdc_types::{GenerationTimestamp, StreamID};
+use anyhow;
+use async_trait::async_trait;
+use futures::future::RemoteHandle;
+use futures::FutureExt;
+use scylla::frame::response::result::CqlValue::Timestamp;
+use scylla::prepared_statement::PreparedStatement;
+use scylla::Session;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::sleep;
+
+/// Alias representing timestamp.
+/// (Last read timestamp, one of stream ids tracked by stream reader, generation of stream)
+pub type Checkpoint = (Duration, StreamID, GenerationTimestamp);
+
+/// Starts creating checkpoints starting from now.
+/// Spawns a new task that periodically creates new checkpoint.
+/// Receives value of the next time window via channel.
+/// We are using tokio::sync::watch channel, because we only need the latest value.
+/// Returns handle to check on this newly created task.
+pub fn start_making_checkpoints(
+    checkpoint_maker: Arc<dyn CDCCheckpointMaker>,
+    receiver: tokio::sync::watch::Receiver<Checkpoint>,
+    saving_period: Duration,
+) -> RemoteHandle<anyhow::Result<()>> {
+    let (fut, handle) = async move {
+        loop {
+            let mut checkpoint = receiver.borrow().clone();
+
+            // We are given timestamp of the beginning of the next read window,
+            // so we have to create a checkpoint with lower timestamp.
+            if checkpoint.0 > Duration::from_millis(1) {
+                checkpoint.0 -= Duration::from_millis(1);
+                checkpoint_maker.make_checkpoint(checkpoint).await?;
+            }
+
+            sleep(saving_period).await;
+        }
+    }
+    .remote_handle();
+
+    tokio::task::spawn(fut);
+
+    handle
+}
+
+/// Customizable trait responsible for creating checkpoints.
+#[async_trait]
+pub trait CDCCheckpointMaker: Send + Sync {
+    /// Creates new checkpoint with given timestamp.
+    async fn make_checkpoint(&self, checkpoint: Checkpoint) -> anyhow::Result<()>;
+    /// Loads last seen generation from the checkpoint table.
+    async fn load_last_generation(&self) -> anyhow::Result<Option<GenerationTimestamp>>;
+    /// Loads last seen timestamp for given stream id from the checkpoint table.
+    async fn load_last_checkpoint(
+        &self,
+        stream_id: &StreamID,
+    ) -> anyhow::Result<Option<chrono::Duration>>;
+}
+
+/// Default implementation for CDCCheckpointMaker trait.
+pub struct CDCDefaultCheckpointMaker {
+    session: Arc<Session>,
+    keyspace: String,
+    checkpoint_tablename: String,
+    prepared_stmt: PreparedStatement,
+}
+
+impl CDCDefaultCheckpointMaker {
+    /// Creates new CDCDefaultCheckpointMaker.
+    /// If given None value as table name,
+    /// it uses default value "cdc_checkpoint".
```

Comment:
The value you've described here is different from the one you're actually using.

Or maybe I am wrong and Scylla doesn't distinguish between upper and lower case but it is still confusing to use both ways here.

Reply:
Also ->
```
`"cdc_checkpoint"`
```